feat(service-worker): expose RegistrationOptions token to allow for runtime configuration

deebloo opened the pull request:
Expose the RegistrationOptions DI token to allow for runtime configuration of ServiceWorkerModule. At the moment there is no way to mark ServiceWorkerModule as enabled at runtime since the configuration values are set at build time.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
At the moment there is no way to mark the sw module as enabled at runtime.  For example if you use global variables for environment variables.

```TS

@NgModule({
  imports: [
    // this does not work since the configuration happens at compile time.
    ServiceWorkerModule.register('/sw/script.js', { enabled: window['isProdMode'] })
  ]
])
class AppModule {}
```

Issue Number: #22796


## What is the new behavior?
Exposing this token will allow folks to configure the sw with useFactory making runtime config possible.

```TS
export function swConfigFactory() {
  return { enabled: window['isProdMode'] }
}

@NgModule({
  imports: [ServiceWorkerModule.register('/sw/script.js')],
  providers: [
    { provide: RegistrationOptions, useFactory: swConfigFactory }
  ]
])
class AppModule {}
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
feat(service-worker): expose `SwRegistrationOptions` token to allow runtime config

Previously, the ServiceWorker registration options should be defined as
an object literal (in order for them to be compatible with Ahead-of-Time
compilation), thus making it impossible to base the ServiceWorker
behavior on runtime conditions.
This commit allows specifying the registration options using a regular
provider, which means that it can take advantage of the `useFactory`
option to determine the config at runtime, while still remaining
compatible with AoT compilation.

gkalpak commented:
Related to #22796.

deebloo commented:
must not have a good job looking for an issue. I noticed that the test is failing because of public API restrictions. I was wondering the correct way going about updating those type def files. whether by hand or some tool?

gkalpak commented:
There is bazel command you can run to update the "golden files". It is mentioned in the CI logs:

```
bazel run //tools/public_api_guard:service-worker_service-worker_api_bin.accept
```

(But for small changes it is usually easier to manually update the corresponding file, [service-worker.d.ts][1] in this case.)

[1]: tools/public_api_guard/service-worker/service-worker.d.ts

gkalpak reviewed (changes requested):
What is the motivation for exporting this? Can you add it to the commit message (and PR description)?

If we are going to do it, then we at least need some docs :wink:

deebloo commented:
@gkalpak i updated the PR and the public_api_guard for service-worker

gkalpak commented:
https://github.com/angular/angular/pull/26002#pullrequestreview-156673828

deebloo commented:
@gkalpak I believe I have everything in order. If not then I am missing the point and need more specific direction to penetrate my thick skull 😆

gkalpak commented:
@deebloo, can you give an example of a usecase that was not possible before and is possible with this change?

deebloo commented:
yes. ill update the PR to show what could NOT be done before

deebloo commented:
done. updated with current behavior and better example for new behavior

mary-poppins commented:
You can preview 8632c5d at https://pr26002-8632c5d.ngbuilds.io/.
You can preview 15e19bf at https://pr26002-15e19bf.ngbuilds.io/.
You can preview e6e12d6 at https://pr26002-e6e12d6.ngbuilds.io/.

gkalpak reviewed (approved):
Thx for the example. What we need:

- Squash the commits into one (that follows [our guidelines][1]).
- Add some API docs for `RegistrationOptions` (see #23138 for example).

##
It would be rad (but not necessary for this PR) if you wanted to add [an example][2] that shows how to use this feature (and has tests to ensure it doesn't break in the future). Let me know if you are interested :wink: 

[1]: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
[2]: https://github.com/angular/angular/tree/master/packages/examples

mary-poppins commented:
You can preview 3308537 at https://pr26002-3308537.ngbuilds.io/.

deebloo commented:
@gkalpak id be happy to add

mary-poppins commented:
You can preview a03f9e4 at https://pr26002-a03f9e4.ngbuilds.io/.

deebloo commented:
@gkalpak I updated the unit tests but not really sure the best way to add an example. Let me know if you think this is good enough

mary-poppins commented:
You can preview 89c5338 at https://pr26002-89c5338.ngbuilds.io/.

gkalpak committed:
test(service-worker): expand `SwRegistrationOptions` tests and move to separate file

gkalpak committed:
docs(service-worker): improve `SwRegistrationOptions` docs and add example

mary-poppins commented:
You can preview a7cbe62 at https://pr26002-a7cbe62.ngbuilds.io/.

gkalpak commented:
Thx for adding the unit tests, @deebloo. I moved them to a separate file (since `comm_spec.ts` is more specifically about the communication between the app and the SW instance) and expanded them a bit. (I hope you don't mind :grin:)

I also realized that adding the example app (with e2e tests) would be a little involved for a first-time contribution :grin:
So, I added a couple of commits with the example app, some nasic e2e tests and a code-snippet to show the example in the docs. Let me know what you think.

mary-poppins commented:
You can preview 82ff361 at https://pr26002-82ff361.ngbuilds.io/.

mary-poppins commented:
You can preview 31c148c at https://pr26002-31c148c.ngbuilds.io/.

deebloo commented:
@gkalpak thanks I appreciate all of the help here. 😄

mary-poppins commented:
You can preview d2e0409 at https://pr26002-d2e0409.ngbuilds.io/.

mary-poppins commented:
You can preview 442626c at https://pr26002-442626c.ngbuilds.io/.

deebloo commented:
@gkalpak really thank you. I am already more comfortable in the codebase. perhaps there will be more PRs to come!

gkalpak commented:
Thank _you_ for putting in all the work. I know it can be an intimidating experience at first, but you did great :smiley:
Looking forward to the next PR 😉

mary-poppins commented:
You can preview 6b749e7 at https://pr26002-6b749e7.ngbuilds.io/.

gkalpak commented:
Rebased on master.

mary-poppins commented:
You can preview 7054dda at https://pr26002-7054dda.ngbuilds.io/.

gkalpak commented:
Oh, the `test_ivy_jit` job is failing now 😞
Investigating...

gkalpak commented:
Since the test passes in the `test` CircleCI job and fails in the `test_ivy_jit` job, I guess it is an Ivy issue.
I assume the correct remedy is tagging `packages/service-worker/test` as `fixme-ivy-jit`, which will exclude it from `test_ivy_jit`.

mary-poppins commented:
You can preview c56b36e at https://pr26002-c56b36e.ngbuilds.io/.

Splaktar reviewed (approved):
LGTM!

gkalpak commented:
Don't worry about it, @deebloo. There is nothing to fix :grin:
It is because I pushed some commits to your PR ([this comment][1] pretty much explains what's going on).

A Googler will take care of it before merging.

[1]: https://github.com/angular/angular/pull/26002#issuecomment-424703653

mary-poppins commented:
You can preview 7ce9134 at https://pr26002-7ce9134.ngbuilds.io/.

mary-poppins commented:
You can preview 41a8e9e at https://pr26002-41a8e9e.ngbuilds.io/.

znikola commented:
Hi,
I can see the PR is ready, is there anything left to do?

deebloo commented:
I don't think so I am just waiting now.

gkalpak commented:
It is just pending approval from @IgorMinar (but pushing ivy out the door is a higher priority at the moment, so things move slowly on other fronts) :smiley:

Splaktar commented:
Looks like a rebase is needed. No rush though.

deebloo commented:
@Splaktar yup yup will handle today

mary-poppins commented:
You can preview e217161 at https://pr26002-e217161.ngbuilds.io/.

IgorMinar reviewed (changes requested):
thanks for this PR.. and sorry for the terribly long delay...

can you please look at my feedback at https://github.com/angular/angular/pull/21842#pullrequestreview-216959549 ? 

That PR is older and seems to be implementing a feature that is a superset of this one. Could we converge these two PRs? @gkalpak could you help facilitate that? thanks

gkalpak commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -1,3 +1,8 @@
+export declare abstract class RegistrationOptions {
```
In https://github.com/angular/angular/pull/21842#discussion_r267528168, @IgorMinar wrote:

this API doesn't look quite right to me because:

1. we now have two ways of enabling the sw (one via provider and one via `.register` method - and these two APIs don't have the same signature so it does look like two separate things. Why not type `opts` below as `RegistrationOptions`?
2. as a DI token all you'll see in the bootstrap/ngmodule code is `RegistrationOptions` and it won't be obvious that this is SW options, if anything we should prefix this with Sw as we've done with other apis under @angular/service-worker

gkalpak commented:
@IgorMinar: I had rebased #21842 on top of this PR (see https://github.com/angular/angular/pull/21842#issuecomment-475038347).
This needs to be merged first. I'll work with @deebloo on getting this in a mergeable state.

deebloo commented:
thanks @gkalpak let me know what I can do

gkalpak commented:
Our infrastructure for running example tests has changed significantly (hence the rebase conflicts you got :wink:), so I need to update those examples/tests.
I can take it from here. Thx, @deebloo!

gkalpak committed:
refactor(service-worker): rename `RegistrationOptions` to `SwRegistrationOptions`

This is in preparation of making `RegistrationOptions` part of the
public API (in a subsequent commit).

deebloo commented:
oh cool thanks @gkalpak ill keep and eye on the PR so I can see what is different

gkalpak commented:
@deebloo, I think it's ready. Please, take a look and let me know if you spot anything dubious :smiley:

More or less here is what I have done:
- Rebased on master.
- Added a commit (15b8cf485) to rename `RegistrationOptions` to `SwRegistrationOptions` (as suggested in https://github.com/angular/angular/pull/21842#pullrequestreview-216959549).
- Used `SwRegistrationOptions` in `ServiceWorkerModule.register()` (as suggested in https://github.com/angular/angular/pull/21842#pullrequestreview-216959549).
- Removed unnecessary commits (for config files that have since been removed).
- Adapted the example to use the new bazel-based setup (1f8c38dd9).

mary-poppins commented:
You can preview 1f8c38d at https://pr26002-1f8c38d.ngbuilds.io/.

deebloo commented:
@gkalpak everything looks like it makes sense. I appreciate all the help getting this done.

gkalpak commented:
@IgorMinar, PTAL.

glebmachine commented:
Please, publish

It's very important due to #26500 bug!
Service worker in Safari is totally broken

IgorMinar reviewed (approved):
lgtm! thanks for getting this into a mergeable state and iterating on the PR

@gkalpak thanks for helping out! ❤️

gkalpak commented:
Thx for working in this, @deebloo ❤️
I am going to close this and include the commit in #21842 (per https://github.com/angular/angular/pull/21842#pullrequestreview-230459821).

deebloo commented:
Of course! I do really appreciate all the assistance.

znikola commented:
Hi, is this going to be in Angular 7 as well?

Splaktar commented:
@znikola I don't believe so.
